### PR TITLE
Json output through languaget interface

### PR DESCRIPTION
**Pull request by peterschrammel**
This removes the dependency of `util` on `langapi`.

~Depends on https://github.com/diffblue/cbmc/pull/2111~ merged

~Depends on https://github.com/diffblue/cbmc/pull/2115~ merged

### CODE REVIEW

**Comment by kroening** [comment]
Does source_locationt really need language-specific conversion?

**Comment by peterschrammel** [comment]
It has the bytecode index as an attribute, which will eventually be moved to a `java_source_locationt`, derived from `source_locationt`.

File: src/goto-programs/loop_ids.cpp [hunk 1]
```diff
@@ -110,6 +110,7 @@ void show_loop_ids(
       forall_goto_functions(it, goto_functions)
         show_loop_ids_json(ui, it->second.body, loops);
 
+      // TODO: this needs clean up
```
Comment (peterschrammel) [inline src/goto-programs/loop_ids.cpp:113]: Needs separate fixing (https://github.com/diffblue/cbmc/issues/1757#issuecomment-383376699)

File: src/analyses/dependence_graph.cpp [hunk 2]
```diff
@@ -272,6 +272,7 @@ jsont dep_graph_domaint::output_json(
   const namespacet &ns) const
 {
   json_arrayt graph;
+  json_exprt json;
```
Comment (peterschrammel) [inline src/analyses/dependence_graph.cpp:275]: Might need separate fixing if language-specific output is desired (https://github.com/diffblue/cbmc/issues/1757#issuecomment-383376699)

File: src/goto-programs/loop_ids.cpp [hunk 3]
```diff
@@ -87,7 +87,7 @@ void show_loop_ids_json(
 
       json_objectt &loop=loops.push_back().make_object();
       loop["name"]=json_stringt(id);
-      loop["sourceLocation"]=json(it->source_location);
+      loop["sourceLocation"] = json_exprt()(it->source_location);
```
Comment (peterschrammel) [inline src/goto-programs/loop_ids.cpp:90]: Needs separate fixing (https://github.com/diffblue/cbmc/issues/1757#issuecomment-383376699)

**Comment by tautschnig** [comment]
May I suggest that the work done here is used as a basis for multiple PRs? It looks like a series of independent bug-fixes are happening here.

**Comment by peterschrammel** [comment]
@tautschnig, I can split up the first 7 commits into a few independent PRs. The last 3 commits that contain the big chunk of work, however, depend on everything before.

**Comment by tautschnig** [comment]
Maybe it's just me, but I found it really hard to make sense of the changes when looking at the combined diff. It may be easier when doing commit-by-commit, but meaningful PRs might help get this merged very quickly.

**Comment by peterschrammel** [comment]
@tautschnig, this is now rebased on develop (two previous PRs have been merged). I've split the remaining commits to make them more digestible, ensuring that all commits build. Please re-review.

File: src/goto-programs/json_goto_trace.h [hunk 4]
```diff
@@ -21,6 +21,9 @@ Date: November 2005
 #include <util/json_expr.h>
 #include <util/invariant.h>
 
+#include <langapi/mode.h>
```
Comment (tautschnig) [inline src/goto-programs/json_goto_trace.h]: Why is this file required here?

**Review by tautschnig — APPROVED** [review]
Looks ok to me.

**Comment by peterschrammel** [comment]
Labelling "do not merge" because TG bump is not passing yet.

**Comment by smowton** [comment]
Similar to the other PR, I suggest distinguishing `languaget`, which is a pretty heavy-weight parser / converter class, from a lighter (probably singleton) `language_infot` used to do stateless conversions like this. Constructing and then destroying the whole Java parsing machinery every time I want to format an expression seems like it would be costly.

**Comment by thk123** [comment]
Due to above comment will hold of reviewing this - please let me know if you'd like it reviewed as is

**Comment by TGWDB** [comment]
Closing this as it appears there is no progress on this at the moment. Also this appears to relate to https://github.com/diffblue/cbmc/pull/2157 that may be required to progress well here.

If you believe this is erroneous, please reopen this PR.
